smartsim/_core/_cli/test.py

```diff
@@ -0,0 +1,215 @@
+        with _VerificationTempDir() as temp_dir:
+            test_install(
+                location=temp_dir,
+                port=None,  # TODO: allow users to pass as arg `--port`?
+                with_tf="tensorflow" in backends,
+                with_pt="torch" in backends,
+                with_onnx="onnxruntime" in backends,
+            )
+    except Exception as e:
+        logger.error(
+            "SmartSim failed to run a simple experiment!\n"
+            f"Experiment failed due to the following exception:\n{e}\n\n"
+            f"Output files are available at `{temp_dir}`"
+        )
+        return 2
+    return 0
+
+
+def test_install(
+    location: str,
+    port: t.Optional[int],
+    with_tf: bool,
+    with_pt: bool,
+    with_onnx: bool,
+) -> None:
+    exp = Experiment("TestExperiment", exp_path=location, launcher="local")
+    port = port or _find_free_port()
+    with _make_managed_local_orc(exp, port) as client:
+        logger.info("Verifying Tensor Transfer")
+        client.put_tensor("plain-tensor", np.ones((1, 1, 3, 3)))
+        client.get_tensor("plain-tensor")
+        if with_tf:
+            logger.info("Verifying TensorFlow Backend")
+            _test_tf_install(client)
+        if with_pt:
+            logger.info("Verifying Torch Backend")
+            _test_torch_install(client)
+        if with_onnx:
+            logger.info("Verifying ONNX Backend")
+            _test_onnx_install(client)
+
+
+@contextmanager
+def _make_managed_local_orc(
+    exp: Experiment, port: int
+) -> t.Generator[Client, None, None]:
+    """Context managed orc that will be stopped if an exception is raised"""
+    orc = exp.create_database(db_nodes=1, interface="lo", port=port)
+    exp.generate(orc)
+    exp.start(orc)
+    try:
+        (client_addr,) = orc.get_address()
+        yield Client(address=client_addr, cluster=False)
+    finally:
+        exp.stop(orc)
+
+
+def _find_free_port() -> int:
+    """A 'good enough' way to find an open port to bind to"""
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
+        sock.bind(("0.0.0.0", 0))
+        _, port = sock.getsockname()
+        return t.cast(int, port)
+
+
+def _test_tf_install(client: Client) -> None:
+    from tensorflow import keras
+
+    from smartsim.ml.tf import serialize_model
+
+    fcn = keras.Sequential(
+        layers=[
+            keras.layers.InputLayer(input_shape=(28, 28), name="input"),
+            keras.layers.Flatten(input_shape=(28, 28), name="flatten"),
+            keras.layers.Dense(128, activation="relu", name="dense"),
+            keras.layers.Dense(10, activation="softmax", name="output"),
+        ],
+        name="FullyConnectedNetwork",
+    )
+    fcn.compile(
+        optimizer="adam", loss="sparse_categorical_crossentropy", metrics=["accuracy"]
+    )
+    model, inputs, outputs = serialize_model(fcn)
+
+    client.set_model(
+        "keras-fcn", model, "TF", device="CPU", inputs=inputs, outputs=outputs
```

Comment:
For all these backend tests, let's add an optional flag to `validate`, `--device` to specify whether we're testing CPU or GPU. GPU is almost always where we run into trouble

Reply:
I'm a little worried about feature creep as this isn't directly related to moving the torch deps. Should we put this into this PR? Would you mind if I kicked this suggestion out into the [`--port` flag](https://github.com/CrayLabs/SmartSim/pull/336/files/ac4734676050a5866d9b1a7013e66a1993d8277f#diff-0f39699c06b8555c21a533e5b264ca4ffc290d50284859a3e1710d1a89f1304c) ticket previously surfaced?

Reply:
NVM, I ended up throwing something together in d9316f595f099a8ee79c0dd042bdca8acaa7532d and e1720e1d5ceb876ba2ccc2a602e8d20220122d39. You should now be able to validate against a GPU